### PR TITLE
[ENH] Dataset object implementation

### PR DESCRIPTION
**Pull request by felipeangelimvieira**
Implementation of dataset objects

#### Reference Issues/PRs
Related to #4333

#### What does this implement/fix? Explain your changes.

This PR builds on draft #4333 by @fkiraly to create a dataset object.

#### What should a reviewer focus on?

This PR is still a work in progress and is intended to facilitate discussion. Some modifications have been made compared to the original draft. 

**Changes implemented so far:**

- A mixin, `_DatasetFromLoaderMixin`, has been created to manage datasets based on loader functions for both classification and forecasting tasks.
- Forecasting datasets and additional classification datasets have been implemented.
- Test classes have been added for both use cases.

**Todos and discussion points:**

1. **Tags and redundancy:**  
   - I am uncertain about the current tags' design. For example, there seems to be redundancy between `n_panels` and `is_one_panel`.  
   - Additionally, while implementing, I found the number of tags confusing (e.g., deciding which tags to use for classification versus forecasting).  
   - Would it be helpful to have a dict-like tag object to assist users in filling tags and reduce redundancy? For instance, the value of `is_one_panel` could depend on `n_panels`.

2. **`Bunch`-like object:**  
   - I did not implement a `Bunch` object in the style of sklearn. However, I believe metadata returned by sklearn loaders (e.g., [this example](https://scikit-learn.org/1.5/modules/generated/sklearn.datasets.load_iris.html#sklearn.datasets.load_iris)) could be stored in the dataset object, similar to how fitted parameters are handled.

3. **Handling test/train splits:**  
   - Currently, the classes raise an exception if a user attempts to retrieve a train/test split from a dataset that doesn’t include one. I believe users should handle splitting such datasets themselves as needed.

4. **Exogenous variable handling:**  
   - For forecasting datasets without exogenous variables, calling `load("X")` currently returns `None`. I believe this approach is preferable since users wouldn’t need to adapt their code when switching between datasets with or without exogenous variables.

I would love to obtain feedback and opinions regarding the current state.

### CODE REVIEW

**Comment by fkiraly** [comment]
About the discussion points:

1. **Tags and redundancy:**  

> I am uncertain about the current tags' design. For example, there seems to be redundancy between `n_panels` and `is_one_panel`.  

This mirrors some of the tags returned by `check_mtype`, which also has some tags that imply each other. I think it makes sense to try to apply all tags that `check_mtype` returns.

> Additionally, while implementing, I found the number of tags confusing (e.g., deciding which tags to use for classification versus forecasting).  

Yes, this is also something I have struggled with, since what a dataset is intended for is not always intrinsic to the data.

> Would it be helpful to have a dict-like tag object to assist users in filling tags and reduce redundancy? For instance, the value of `is_one_panel` could depend on `n_panels`.

Can you explain this? Why not a `dict`?

2. **`Bunch`-like object:**  

> I did not implement a `Bunch` object in the style of sklearn. However, I believe metadata returned by sklearn loaders (e.g., [this example](https://scikit-learn.org/1.5/modules/generated/sklearn.datasets.load_iris.html#sklearn.datasets.load_iris)) could be stored in the dataset object, similar to how fitted parameters are handled.

Yes, that would make sense, although I do not have a strong opinion either way.


3. **Handling test/train splits:**  

> Currently, the classes raise an exception if a user attempts to retrieve a train/test split from a dataset that doesn’t include one. I believe users should handle splitting such datasets themselves as needed.

I disagree with this - many datasets have a "predefined split" to them, that people consider part of the data set.

There is a semantic distinction between simply "data", and a data-plus-split-for-a-specific-task. I have struggled design-wise with whether to distinguish these, and if yes, how.

4. **Exogenous variable handling:**  

> For forecasting datasets without exogenous variables, calling `load("X")` currently returns `None`. I believe this approach is preferable since users wouldn’t need to adapt their code when switching between datasets with or without exogenous variables.

Agreed.

Although as above, this makes sense only if there is a clear forecasting purpose attached to the data. Otherwise it is not clear what "X" would be.

**Comment by felipeangelimvieira** [comment]
Thank you for the reply Franz

> > Would it be helpful to have a dict-like tag object to assist users in filling tags and reduce redundancy? For instance, the value of is_one_panel could depend on n_panels.
> 
> Can you explain this? Why not a dict?

This is just an idea but could work somewhat like this:

```python
class BaseDependentDict(dict):
    _dependent_keys = {}

    def __getitem__(self, key):
        if key in self._dependent_keys:
            return self._dependent_keys[key](self)
        return super().__getitem__(key)

    def __iter__(self):
        return iter(set(super().keys()).union(self._dependent_keys.keys()))

    def keys(self):
        return set(super().keys()).union(self._dependent_keys.keys())

    def items(self):
        return ((key, self[key]) for key in self.keys())

    def values(self):
        return (self[key] for self.keys())

    def get(self, key, default=None):
        if key in self._dependent_keys or key in self:
            return self[key]
        return default

    def __contains__(self, key):
        return key in self._dependent_keys or super().__contains__(key)

class DatasetTagDict(BaseDependentDict):
    _dependent_keys = {
        "is_one_panel": lambda self: self["n_panels"] == 1,
        # Additional dependent keys can be defined here
    }
```

The objective here is to avoid manual errors, such as conflicting values. While I'm suggesting this idea, it isn’t essential for this PR. It might be something to consider for future improvements.

> I disagree with this - many datasets have a "predefined split" to them, that people consider part of the data set.
> 
> There is a semantic distinction between simply "data", and a data-plus-split-for-a-specific-task. I have struggled design-wise with whether to distinguish these, and if yes, how.

Franz, just to confirm: do you prefer that we suggest default train-test splits for datasets that don’t have an "official" split? For example, the `load_airline` function currently doesn't include an official test set. Should the airline class implement a default split?

> Although as above, this makes sense only if there is a clear forecasting purpose attached to the data. Otherwise it is not clear what "X" would be.

Hmm, I hadn’t considered this. The `BaseDataset` would ideally be general enough to support tasks beyond time series-related ones?

**Comment by fkiraly** [comment]
> The objective here is to avoid manual errors, such as conflicting values. It isn’t essential for this PR, might be something to consider for future improvements, or just keep as-is.

I think this is "overkill" - I would instead test it externally, i.e., automated tests check that the fields are consistent. That way, the core structures are simple and standard, while consistency is still ensured in a systematic fashion.

> Franz, just to confirm: do you prefer that we suggest default train-test splits for datasets that don’t have an "official" split? For example, the `load_airline` function currently doesn't include an official test set. Should the airline class implement a default split?

No, there was no concrete suggestion implied - I was trying to say, in our mental model we need to account both for abstract datasets that *do* contain a default split, and ones that *do not*. We can still actively make the decision to not cover one of the two cases, but that was not what I was suggesting (or any alternative).

The way I tried to account for this in the original design was to allow different "fields" in the dataset - the "no predefined split" case would have `X` and `y` only, the "predefined split" caes would also have `X_train`, `X_test`, `y_train`, `y_test`.

**Comment by felipeangelimvieira** [comment]
Just added the changes as described in sktime/enhancement-proposals/pull/37.

@fkiraly, I've also added a new class parameter in BaseObject `_MIN_TEST_PARAMS_COVERAGE`, with the idea that datasets do not need 2 test sets (some of them do not have init parameters at all). What do you think of that?

**Comment by felipeangelimvieira** [comment]
I could also just skip the test for minimum param coverage if the object is a Dataset

**Comment by fkiraly** [comment]
> @fkiraly, I've also added a new class parameter in BaseObject `_MIN_TEST_PARAMS_COVERAGE`, with the idea that datasets do not need 2 test sets (some of them do not have init parameters at all). What do you think of that?

Hm, I think it is unnecessary - the test for test param coverage already skips all objects that have no parameters at all, afaik.

**Review by fkiraly — COMMENTED** [review]
Is this ready for review? Monumental!

**Review by fkiraly — CHANGES_REQUESTED** [review]
I think this PR is brilliant!

Blocking, for discussion:

* could we call the new types `dataset_classification` instead of `classification_dataset`? That way they are lexicographically close and follow the naming pattern `[type]_[subtype]`
    * also, should the more specialized types also be datasets? That is, should the `object_type` not be the list `["dataset_classification", "dataset"]`?
* the change to `test_get_test_params_coverage` seems unnecessary - does it not already allow only a single parameter set when the object has no parameters?
    * or is the intention to skip the test entirely to avoid data downloads etc? Clarification would be appreciated

Non-blocking:

* in the design, you have separated "dataset for classification" etc. I assume you have thought carefully that these cannot be dealt with a single base class?

**Comment by felipeangelimvieira** [comment]
> Is this ready for review? Monumental!
Yes! I was trying to figure out the CI errors. The macOS is known, I was also investigating the one of Cython estimators.

> could we call the new types dataset_classification instead of classification_dataset? That way they are lexicographically close and follow the naming pattern [type]_[subtype]
> also, should the more specialized types also be datasets? That is, should the object_type not be the list ["dataset_classification", "dataset"]?

For sure, I didn't know we could set object_type to a list.

> the change to test_get_test_params_coverage seems unnecessary - does it not already allow only a single parameter set when the object has no parameters?
> or is the intention to skip the test entirely to avoid data downloads etc? Clarification would be appreciated
> if this is about "parameters that every class has" like return_mtype, you can add them to the tag "reserved_params", and then they do not count for the test.

This test was raising an error because there weren't the minimum of 2 test parameters combinations. Is there another way of skipping it? Returning None in `get_test_params`?

> Non-blocking:
> 
> in the design, you have separated "dataset for classification" etc. I assume you have thought carefully that these cannot be dealt with a single base class?

At first, I've tried implementing all of them with a single base class. However, their tags need to be different, and also the tests. But now that you've asked, I was thinking if this could also be handled by Mixins. What would you suggest?

**Comment by fkiraly** [comment]
> For sure, I didn't know we could set object_type to a list.

It is actually in the docs now!
You might not have noticed, but we have documented the tags (at least, the most important ones)
https://www.sktime.net/en/latest/api_reference/auto_generated/sktime.registry._tags.object_type.html

> This test was raising an error because there weren't the minimum of 2 test parameters combinations. Is there another way of skipping it?

This should not happen if the class has no parameters - could you revert or make another branch to test whether this is a bug? I mean, establish the condition where it was failing, so I can debug it. Can e this PR or separate PR/branch. Would be much appreciated.

> At first, I've tried implementing all of them with a single base class. However, their tags need to be different, and also the tests. But now that you've asked, I was thinking if this could also be handled by Mixins. What would you suggest? I could update it.

Mixins sound messy, especially with the tests. Although the tests could collect by tags. The risk if we leave it is that we cannot easily change it after the release - and if we change it now, you might run into a frustrating corner where things no longer work... up to you. Perhaps if you try it, separate branch, to avoid making a revert difficult if it ends up being frustrating?

**Comment by felipeangelimvieira** [comment]
Ok! I will try using the same base class, and different object tags, to be able to filter in the unit tests

**Comment by felipeangelimvieira** [comment]
@fkiraly this morning I tested an implementation based on a single class. It worked, but I have mixed feelings about having a lot of tags in `BaseDataset` class that may not be used by some of the children. Example:


```python
class BaseDataset(BaseObject):
    """Base class for datasets."""

    # default tag values - these typically make the "safest" assumption

_tags = {
        "object_type": "dataset",  # type of object
        "name": None,  # The dataset unique name
        "python_dependencies": None,  # python dependencies required to load the dataset
        "python_version": None,  # python version required to load the dataset
        ## General tags
        "n_splits": 0,  # Number of cross-validation splits, if any.
        "is_univariate": True,
        "n_instances": None,
        "n_instances_train": None,
        "n_instances_test": None,
        ## Classification specific tags
        "n_classes": 2,
        ## Forecasting specific tags
        "is_equally_spaced": True,
        "has_nans": False,
        "has_exogenous": False,
        "n_timepoints": None,
        "n_timepoints_train": None,
        "n_timepoints_test": None,
        "frequency": "M",
        "n_dimensions": 1,
        "is_one_panel": True,
        "n_panels": 1,
        "n_hierarchy_levels": 0,
        "is_one_series": True,
        ## Regression specific tags
        # Regression datasets only have the general tags
    }
```

This generates two problems: first, if we search by tags exclusive to forecasting datasets, some classification datasets may also be returned; second, if we want to create new types of datasets in sktime, such as anomaly detection, we would have to add even more tags to the base class. 

Having a minimal `BaseDataset` class may also be useful for 2nd and 3rd party libraries, such as in `Prophetverse`, where we could have a `MarketingMixModelingDataset`, without inheriting classification/regression dataset tags.

A potential solution, while still trying to keep a single base class, is not adding these default tag values to `BaseDataset`, and adding only to the children, based on their application (forecasting/classification).  However, this would remove the safest default values, and may also be more difficult to maintain.

So I still believe that having classes related to each timeseries problems is better in terms of cognitive load. In the current state, each base class serves to set default tag values - potentially in the future adding checks between public and private load methods?

Am I missing something? I would appreciate some feedback on that

**Comment by fkiraly** [comment]
> second, if we want to create new types of datasets in sktime, such as anomaly detection, we would have to add even more tags to the base class.

I see - I think this is a very strong argument - extensibility would be compromised if we merge the classes.

I am now convinced that having separate classes is a better way to proceed.

**Review by fkiraly — CHANGES_REQUESTED** [review]
I think we should merge as-is, except the stepout with the min_parameters, which imo is not needed (see above).

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great!

I have fixed the issue with the failing test params test, this required to add the `reserved_params` tag.

Some other things I noticed while going through:

* for documentation of the data formats, we should no longer link to python objects but directly to the docs - I have done this in the classification datasets, as example.
* some docstrings give contradictory default mtypes - in the first line, and then in the body. What is the default return type? Would be great if you could check. I left one instance of contradictcion in the classification docstring.
* If you have docstrings with indentations like you had, they do not render correctly. Also, newlines get removed unless you have two newlines, or bullet points. I have reformatted the classificatoin docstrings as an example.
* should forecasting and regression datasets also have `return_mtype` fields? Currently, only classification has this.

Finally, suggestions for improvement:
* should we add the new datasets somewhere in the API reference for datasets? Would be great, given that we are now able to give each dataset a full page of space, with a complete description.
* I would name the tags and values differently. Object types `dataset_regression` etc, to ensure it appears as a block in lexicographical orderung.
    * the tag `object_type` should be polymorphic, e.g., `["dataset_classification", "dataset"]`. That way, we can use `dataset` as a higher-level type that collects all datasets.
    * Idea: add another tag `task_type` which has *the same values* as estimators, e.g., `"regressor"`, `"forecaster"`. That way, you can collect the applicable estimators via a programmatic database join.

**Comment by fkiraly** [comment]
PS: the failure for `Longley` is real - the test is testing what it is supposed to test.

**Review by fkiraly — APPROVED** [review]
